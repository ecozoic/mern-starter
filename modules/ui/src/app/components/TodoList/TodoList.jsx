import React from 'react';
import PropTypes from 'prop-types';

import { todoShape } from '../../models';
import TodoComponent from '../Todo';

import styles from './TodoList.scss';

/**
 * Renders a list of todos
 * @param {Object} props - component props
 * @param {Object[]} props.todos - array of todos
 * @param {Function} props.onTodoClick - function to call when todo is clicked
 */
const TodoList = ({ todos, onTodoClick }) => (
  <ul className={styles.todoList}>
    {todos.map(todo => (
      <TodoComponent
        key={todo._id}
        onClick={() => onTodoClick(todo._id)}
        completed={todo.completed}
        text={todo.text}
      />
    ))}
  </ul>
);

TodoList.propTypes = {
  todos: PropTypes.arrayOf(todoShape).isRequired,
  onTodoClick: PropTypes.func.isRequired,
};

export default TodoList;
