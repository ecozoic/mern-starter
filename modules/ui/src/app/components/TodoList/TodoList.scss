.todo-list {
  cursor: pointer;
}

.progress-container {
  padding-top: 20px;
  text-align: center;
}
