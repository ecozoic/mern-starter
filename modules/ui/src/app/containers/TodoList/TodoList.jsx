import { connect } from 'react-redux';

import { toggleTodo } from '../../actions';
import { getTodos } from '../../selectors';
import TodoList from '../../components/TodoList';

const mapStateToProps = state => ({
  todos: getTodos(state),
});

const mapDispatchToProps = dispatch => ({
  onTodoClick: (id) => {
    dispatch(toggleTodo(id));
  },
});

/**
 * React-redux container for TodoList component
 */
const TodoListContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(TodoList);

export default TodoListContainer;
