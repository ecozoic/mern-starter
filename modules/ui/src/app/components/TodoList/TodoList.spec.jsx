import React from 'react';
import _ from 'lodash';
import { shallow } from 'enzyme';

import TodoList from './';

describe('<TodoList />', () => {
  it('renders without crashing', () => {
    const todoList = shallow(
      <TodoList
        todos={[]}
        onTodoClick={_.identity}
      />,
    );

    expect(todoList).toMatchSnapshot();
  });

  it('displays an unordered list', () => {
    const todoList = shallow(
      <TodoList
        todos={[]}
        onTodoClick={_.identity}
      />,
    );

    expect(todoList.find('ul').length).toBe(1);
  });

  it('renders a todo for each todo in props', () => {
    const todos = [
      { id: '1', text: 'Todo 1', completed: false },
      { id: '2', text: 'Todo 2', completed: true },
      { id: '3', text: 'Todo 3', completed: false },
    ];

    const todoList = shallow(
      <TodoList
        todos={todos}
        onTodoClick={_.identity}
      />,
    );

    expect(todoList.find('Todo').length).toBe(3);
  });

  it('renders todo with proper props', () => {
    const todos = [
      { id: '1', text: 'Todo 1', completed: false },
    ];

    const todoList = shallow(
      <TodoList
        todos={todos}
        onTodoClick={_.identity}
      />,
    );

    const todo = todoList.find('Todo');
    const props = todo.props();

    expect(props.completed).toBe(false);
    expect(props.text).toBe('Todo 1');
  });

  it('sets the todo id as the key', () => {
    const todos = [
      { id: '1', text: 'Todo 1', completed: false },
    ];

    const todoList = shallow(
      <TodoList
        todos={todos}
        onTodoClick={_.identity}
      />,
    );

    expect(todoList.find('Todo').key()).toEqual('1');
  });

  it('calls onTodoClick with id of clicked todo', () => {
    const mockFn = jest.fn();

    const todos = [
      { id: '1', text: 'Todo 1', completed: false },
    ];

    const todoList = shallow(
      <TodoList
        todos={todos}
        onTodoClick={mockFn}
      />,
    );

    todoList.find('Todo').simulate('click');

    expect(mockFn).toHaveBeenCalledTimes(1);
    expect(mockFn.mock.calls[0][0]).toBe('1');
  });
});
