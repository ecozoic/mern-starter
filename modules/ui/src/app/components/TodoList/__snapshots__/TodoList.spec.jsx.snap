// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoList /> renders without crashing 1`] = `
<List
  animated={true}
  bulleted={true}
  relaxed={true}
  size="large"
/>
`;
